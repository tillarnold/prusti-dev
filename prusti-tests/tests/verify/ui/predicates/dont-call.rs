#![allow(dead_code)]

use prusti_contracts::*;

#[predicate]
fn pred_id(x: bool) -> bool {
    x
}

fn illegal_use() {
    let _x = pred_id(true);
}

fn illegal_ref(_pred: fn(bool) -> bool) {}

struct Outer;

impl Outer {
    pub fn nested() {
        struct Inner;

        impl Inner {
            #[predicate]
            fn inner_pred(b: bool) -> bool {
                b
            }

            fn illegal() {
                illegal_ref(Self::inner_pred)
            }
        }
    }
}

fn main() {
    illegal_use();
    illegal_ref(pred_id);
}
