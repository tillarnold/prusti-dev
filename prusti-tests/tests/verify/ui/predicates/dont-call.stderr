error: [Prusti: invalid specification] using predicate from non-specification code is not allowed
  --> $DIR/dont-call.rs:11:14
   |
11 |     let _x = pred_id(true);
   |              ^^^^^^^
   |
note: this is a specification-only predicate function
  --> $DIR/dont-call.rs:6:1
   |
6  | / fn pred_id(x: bool) -> bool {
7  | |     x
8  | | }
   | |_^

error: [Prusti: invalid specification] using predicate from non-specification code is not allowed
  --> $DIR/dont-call.rs:29:29
   |
29 |                 illegal_ref(Self::inner_pred)
   |                             ^^^^^^^^^^^^^^^^
   |
note: this is a specification-only predicate function
  --> $DIR/dont-call.rs:24:13
   |
24 | /             fn inner_pred(b: bool) -> bool {
25 | |                 b
26 | |             }
   | |_____________^

error: [Prusti: invalid specification] using predicate from non-specification code is not allowed
  --> $DIR/dont-call.rs:37:17
   |
37 |     illegal_ref(pred_id);
   |                 ^^^^^^^
   |
note: this is a specification-only predicate function
  --> $DIR/dont-call.rs:6:1
   |
6  | / fn pred_id(x: bool) -> bool {
7  | |     x
8  | | }
   | |_^

error: aborting due to 3 previous errors

